### PR TITLE
Remove some allocations

### PR DESCRIPTION
**Pull request by rubdos**
Performance improvements, i.e. #260 / #261

Fixes no specific issue

### Checklist
- [x] My branch is up-to-date with development branch.
- [x] Everything works and tested on latest stable Rust.
- [ ] Coverage and Linting have been applied 

### Current behaviour
smartcore svm is a bit slow

### New expected behaviour
smartcore svm is a bit less slow! About 4% faster training in my very simple 5 features x 159 samples test.

### Change logs

This removes some allocations by reusing vectors and clearing them in iterations.

### CODE REVIEW

**Comment by rubdos** [comment]
Second commit makes training another 14% faster on my test data.

**Comment by Mec-iS** [comment]
thanks a lot, i will look into it as soon as possible. Please run `rustfmt src/*.rs` to avoid the linting test failing.

**Comment by rubdos** [comment]
FWIW, rustfmt does not complain about my changes; those lines are unrelated to this PR. I'll gladly run rustfmt if you still want me to, however.

I currently notice that my accuracy estimation (possibly `train_test_split`) is the slowest part in my test code. I'm playing with KNN now, however (`Model trained in 1921 ms; accuracy 0.9921170101624086 (estimated in 189165 ms)`).

The difference between training and estimation is only `train_test_split` and `ClassificationMetricsOrd::accuracy().get_score`...

**Comment by Mec-iS** [comment]
good job!

Tomorrow I will have some time at the computer to take a closer look to your contribution.

clippy is complaining for unnecessary parenthesis. For the sake of making the tests pass please apply the requested minor changes even if present in other files.

Thanks again for your time spent using Smartcore.

**Comment by morenol** [comment]
I can create a PR fixing clippy in a couple of hours. That is probably related to the new lints added to the stable version released a couple of days ago

**Comment by Mec-iS** [comment]
thanks Luis 👍
obviously we want 100% to stick with clippy standards.

**Comment by morenol** [comment]
PR opened at: https://github.com/smartcorelib/smartcore/pull/263

We can merge that first. Then we can rebase these changes. Hopefully, CI will be green after that

**Comment by morenol** [comment]
@rubdos could you rebase your changes? It seems that there are not conflicts

File: src/svm/svc.rs [hunk 1]
```diff
@@ -322,19 +322,26 @@ impl<'a, TX: Number + RealNumber, TY: Number + Ord, X: Array2<TX> + 'a, Y: Array
...
         }
 
         Ok(y_hat)
     }
 
-    fn predict_for_row(&self, x: Vec<TX>) -> TX {
+    fn predict_for_row(&self, x: &[TX]) -> TX {
         let mut f = self.b.unwrap();
 
+        let xi: Vec<_> = x.iter().map(|e| e.to_f64().unwrap()).collect();
```
Comment (morenol) [inline src/svm/svc.rs:339]: good catch!

**Review by morenol — DISMISSED** [review]
looks good to me. But needs a rebase

**Review by Mec-iS — DISMISSED** [review]
Thanks a lot for this.

If you rebase your branch with current `development` I will merge.

**Comment by rubdos** [comment]
> If you rebase your branch with current `development` I will merge.

Done!

**Review by Mec-iS — APPROVED** [review]
